book/animations.md

```diff
@@ -638,226 +634,329 @@ trigger it once every 2 seconds:
+def style(node, rules):
+    old_style = node.style
 
     # ...
 
     if old_style:
-        try_numeric_animation(node, "opacity",
-            old_style, node.style, tab, is_px=False)
-        try_numeric_animation(node, "width",
-            old_style, node.style, tab, is_px=True)
-```
-
-Implement the `try_numeric_animation` function. It will compare the
-values of a CSS property in the old and new style. If they are different *and*
-there is a CSS transition defined for those properties, an animation will
-begin. Start with a helper method that returns the duration (in units of animation
-frame count) of a transition if it was set, and `None` otherwise. This requires
-parsing the comma-separated `transition` syntax.^[Unfortunately, setting up
-animations tends to have a lot of boilerplate code, so get ready for more code
-than usual. The good news though is that it's all pretty simple to
-understand.]
+        transitions = diff_styles(old_style, node.style)
+```
+
+This `diff_style` function is going to look for all properties that
+are mentioned in the `transition` property and are different between
+the old and the new style. So first, we're going to have to parse the
+`transition` value:
 
 ``` {.python}
-def get_transition(property_name, style):
-    if not "transition" in style:
-        return None
-    transition_items = style["transition"].split(",")
-    found = False
-    for item in transition_items:
-        if property_name == item.split(" ")[0]:
-            found = True
-            break
-    if not found:
-        return None   
-    duration_secs = float(item.split(" ")[1][:-1])
-    return duration_secs / REFRESH_RATE_SEC 
+def parse_transition(value):
+    properties = {}
+    if not value: return properties
+    for item in value.split(","):
+        property, duration = item.split(" ", 1)
+        frames = float(duration[:-1]) / REFRESH_RATE_SEC
+        properties[property] = frames
+    return properties
 ```
 
-Add some code that detects if a transition remains across a style
-recalc, by comparing two style objects---the ones before and after a style
-update for a DOM node:
+Note that this returns a dictionary mapping property names to
+transition durations, measured in frames.
+
+Next, `diff_style` will loop through all of the properties mentioned
+in `transition` and see which ones changed. It returns a dictionary
+containing only the transitioning properties, and mapping each such
+property to its old value, new value, and duration (again in frames):
 
 ``` {.python}
-def try_transition(property_name, node, old_style, new_style):
-    if not get_transition(property_name, old_style):
-        return None
+def diff_styles(old_style, new_style):
+    old_transitions = parse_transition(old_style.get("transition"))
+    new_transitions = parse_transition(new_style.get("transition"))
+
+    transitions = {}
+    for property in old_transitions:
+        if property not in new_transitions: continue
```

Comment:
you forgot old_transitions.

Reply:
We're iterating over old_transitions and then additionally filtering by new_transitions. So we end up with things in the intersection.

Reply:
ah good point